File: taurus-protect-sdk-java/client/src/test/java/com/taurushq/sdk/protect/client/model/DecodeToleranceVectorsTest.java
package com.taurushq.sdk.protect.client.model;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertNotNull;
import static org.junit.jupiter.api.Assertions.assertTrue;
import static org.junit.jupiter.api.Assertions.fail;

import java.io.IOException;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.security.PublicKey;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import com.google.gson.JsonArray;
import com.google.gson.JsonElement;
import com.google.gson.JsonObject;
import com.google.gson.JsonParser;
import com.google.gson.JsonPrimitive;
import com.taurushq.sdk.protect.client.cache.RulesContainerCache;
import com.taurushq.sdk.protect.client.mapper.ApiExceptionMapper;
import com.taurushq.sdk.protect.client.service.AddressService;
import com.taurushq.sdk.protect.client.service.AssetService;
import com.taurushq.sdk.protect.client.service.GovernanceRuleService;
import com.taurushq.sdk.protect.client.service.MultiFactorSignatureService;
import com.taurushq.sdk.protect.client.service.WhitelistedAddressService;
import com.taurushq.sdk.protect.client.testutil.SignedWhitelist;
import com.taurushq.sdk.protect.client.testutil.StubTransport;
import com.taurushq.sdk.protect.openapi.ApiClient;
import com.taurushq.sdk.protect.openapi.JSON;
import com.taurushq.sdk.protect.openapi.model.TgvalidatordGetBalancesReply;
import com.taurushq.sdk.protect.openapi.model.TgvalidatordGetMultiFactorSignatureEntitiesInfoReply;
import com.taurushq.sdk.protect.openapi.model.TgvalidatordMultiFactorSignaturesEntityType;
import com.taurushq.sdk.protect.openapi.model.TgvalidatordStakeAccount;
import com.taurushq.sdk.protect.openapi.model.TgvalidatordStakeAccountType;
import com.taurushq.sdk.protect.openapi.model.TgvalidatordTokenType;
import org.junit.jupiter.api.BeforeAll;
import org.junit.jupiter.api.Test;

/**
 * Cross-SDK decode tolerance.
 *
 * <p>A field or an enum value the client was not generated with must never fail a decode, and
 * the four SDKs must treat it the same way: unknown fields land in the generated model's
 * additional properties and are written back, unknown enum values keep their raw value all the
 * way into the domain model, and caller-supplied values are sent verbatim. The shared vector
 * file is consumed by the Go/Java/Python/TypeScript suites alike.
 */
class DecodeToleranceVectorsTest {

    private static final Map<String, Class<?>> MODELS = new HashMap<>();
    private static final Map<String, Class<?>> ENUMS = new HashMap<>();

    static {
        MODELS.put("tgvalidatordGetBalancesReply", TgvalidatordGetBalancesReply.class);
        MODELS.put("tgvalidatordStakeAccount", TgvalidatordStakeAccount.class);
        MODELS.put("tgvalidatordGetMultiFactorSignatureEntitiesInfoReply",
                TgvalidatordGetMultiFactorSignatureEntitiesInfoReply.class);
        ENUMS.put("tgvalidatordTokenType", TgvalidatordTokenType.class);
        ENUMS.put("tgvalidatordMultiFactorSignaturesEntityType", TgvalidatordMultiFactorSignaturesEntityType.class);
        ENUMS.put("tgvalidatordStakeAccountType", TgvalidatordStakeAccountType.class);
    }

    private static JsonArray vectors;

    @BeforeAll
    static void loadVectors() throws IOException {
        String[] candidates = {
                "../../scripts/resources/decode-tolerance-vectors.json", // from client/
                "../scripts/resources/decode-tolerance-vectors.json",    // from sdk-java/
                "scripts/resources/decode-tolerance-vectors.json",       // from repo root
        };
        for (String candidate : candidates) {
            Path p = Paths.get(candidate).toAbsolutePath().normalize();
            if (Files.exists(p)) {
                String json = new String(Files.readAllBytes(p), StandardCharsets.UTF_8);
                vectors = JsonParser.parseString(json).getAsJsonArray();
                return;
            }
        }
        throw new IOException("Cannot find decode-tolerance-vectors.json. Working directory: "
                + System.getProperty("user.dir"));
    }

    private static List<JsonObject> ofKind(final String kind) {
        List<JsonObject> matching = new ArrayList<>();
        for (JsonElement element : vectors) {
            JsonObject vector = element.getAsJsonObject();
            if (kind.equals(vector.get("kind").getAsString())) {
                matching.add(vector);
            }
        }
        assertFalse(matching.isEmpty(), "no '" + kind + "' vectors");
        return matching;
    }

    @Test
    void everyVectorIsOfAKindThisSuiteRuns() {
        assertTrue(vectors.size() > 0, "shared vectors file is empty");
        for (JsonElement element : vectors) {
            String kind = element.getAsJsonObject().get("kind").getAsString();
            assertTrue(kind.equals("model") || kind.equals("enum") || kind.equals("service"),
                    "unknown vector kind " + kind);
        }
    }

    @Test
    void generatedModelsKeepUnknownFieldsAndRoundTripLosslessly() throws Exception {
        for (JsonObject vector : ofKind("model")) {
            String name = vector.get("name").getAsString();
            Class<?> type = MODELS.get(vector.get("model").getAsString());
            assertNotNull(type, name + ": no generated class mapped for " + vector.get("model"));
            JsonObject expected = vector.getAsJsonObject("expected");
            String wire = vector.get("wire").toString();

            Object decoded;
            try {
                decoded = JSON.getGson().fromJson(wire, type);
            } catch (RuntimeException e) {
                throw new AssertionError(name + ": decode failed: " + e.getMessage(), e);
            }

            Object extras = type.getMethod("getAdditionalProperties").invoke(decoded);
            assertEquals(expected.get("rootAdditionalProperties"), JSON.getGson().toJsonTree(extras),
                    name + ": root additional properties");

            if (expected.get("lossless").getAsBoolean()) {
                assertEquals(JsonParser.parseString(wire), JsonParser.parseString(JSON.serialize(decoded)),
                        name + ": re-serialization is not lossless");
            }
        }
    }

    @Test
    void generatedEnumsKeepUnknownValuesRaw() throws Exception {
        for (JsonObject vector : ofKind("enum")) {
            String name = vector.get("name").getAsString();
            Class<?> type = ENUMS.get(vector.get("enum").getAsString());
            assertNotNull(type, name + ": no generated enum mapped for " + vector.get("enum"));
            JsonObject expected = vector.getAsJsonObject("expected");

            Object decoded;
            try {
                decoded = JSON.getGson().fromJson(new JsonPrimitive(vector.get("wire").getAsString()), type);
            } catch (RuntimeException e) {
                throw new AssertionError(name + ": decode failed: " + e.getMessage(), e);
            }

            assertEquals(expected.get("value").getAsString(), type.getMethod("getValue").invoke(decoded), name);
            boolean known = false;
            for (Object constant : (Object[]) type.getMethod("values").invoke(null)) {
                known |= constant == decoded;
            }
            assertEquals(expected.get("known").getAsBoolean(), known, name + ": known");
        }
    }

    @Test
    void servicesPassUnknownValuesThroughBothWays() throws Exception {
        for (JsonObject vector : ofKind("service")) {
            String name = vector.get("name").getAsString();
            String operation = vector.get("operation").getAsString();
            JsonObject request = vector.getAsJsonObject("request");
            JsonObject expected = vector.getAsJsonObject("expected");
            String reply = vector.get("reply").toString();
            switch (operation) {
                case "getMultiFactorSignatureInfo":
                    getMultiFactorSignatureInfo(name, request, expected, reply);
                    break;
                case "createMultiFactorSignatures":
                    createMultiFactorSignatures(name, request, expected, reply);
                    break;
                case "queryAssetAddresses":
                    queryAssetAddresses(name, request, expected, reply);
                    break;
                default:
                    fail(name + ": unhandled operation " + operation);
            }
        }
    }

    private static void getMultiFactorSignatureInfo(final String name, final JsonObject request,
                                                    final JsonObject expected, final String reply)
            throws Exception {
        StubTransport stub = StubTransport.replying(reply);
        MultiFactorSignatureService service = new MultiFactorSignatureService(stub.client(), new ApiExceptionMapper());

        MultiFactorSignatureInfo info = service.getMultiFactorSignatureInfo(request.get("id").getAsString());

        assertEquals(expected.get("entityType").getAsString(), info.getEntityType().getKind(), name);
    }

    private static void createMultiFactorSignatures(final String name, final JsonObject request,
                                                    final JsonObject expected, final String reply)
            throws Exception {
        StubTransport stub = StubTransport.replying(reply);
        MultiFactorSignatureService service = new MultiFactorSignatureService(stub.client(), new ApiExceptionMapper());
        List<String> ids = new ArrayList<>();
        for (JsonElement id : request.getAsJsonArray("entityIds")) {
            ids.add(id.getAsString());
        }

        service.createMultiFactorSignatures(ids,
                TgvalidatordMultiFactorSignaturesEntityType.fromValue(request.get("entityType").getAsString()));

        assertSentFields(name, expected.getAsJsonObject("requestBody"), stub.only().body());
    }

    private static void queryAssetAddresses(final String name, final JsonObject request,
                                            final JsonObject expected, final String reply)
            throws Exception {
        String assetId = request.get("assetId").getAsString();
        String path = "/api/rest/v2/assets/" + assetId + "/addresses/query";
        StubTransport stub = StubTransport.replying().route(path, r -> reply);
        ApiClient client = stub.client();
        ApiExceptionMapper mapper = new ApiExceptionMapper();
        List<PublicKey> superAdminKeys = new SignedWhitelist().superAdminKeys();
        RulesContainerCache cache = new RulesContainerCache(
                new GovernanceRuleService(client, mapper, superAdminKeys, 1));
        AssetService assets = new AssetService(client, mapper, cache, new AddressService(client, mapper, cache),
                new WhitelistedAddressService(client, mapper, superAdminKeys, 1));

        AssetAddressV2Result result = assets.queryAssetAddresses(assetId,
                request.get("addressType").getAsString(), null, null, null);

        assertEquals(1, stub.requests().size(), name + ": only the holders page may be read");
        assertSentFields(name, expected.getAsJsonObject("requestBody"), stub.requestsTo(path).get(0).body());
        JsonArray rows = expected.getAsJsonArray("rows");
        assertEquals(rows.size(), result.getAddresses().size(), name + ": rows returned");
        for (int i = 0; i < rows.size(); i++) {
            JsonObject row = rows.get(i).getAsJsonObject();
            AssetAddressV2 got = result.getAddresses().get(i);
            assertEquals(row.get("address").getAsString(), got.getAddress(), name);
            assertEquals(row.get("addressType").getAsString(), got.getAddressType(), name);
            assertEquals(row.get("verified").getAsBoolean(), got.isVerified(), name);
        }
        assertEquals(expected.get("excludedCount").getAsInt(), result.getExcludedUnverified().size(),
                name + ": excluded");
    }

    private static void assertSentFields(final String name, final JsonObject expected, final String body) {
        assertNotNull(body, name + ": the request had no body");
        JsonObject sent = JsonParser.parseString(body).getAsJsonObject();
        for (Map.Entry<String, JsonElement> field : expected.entrySet()) {
            assertEquals(field.getValue(), sent.get(field.getKey()), name + ": sent " + field.getKey());
        }
    }
}

File: taurus-protect-sdk-typescript/tests/unit/decode-tolerance-vectors.test.ts
/**
 * Cross-SDK decode tolerance.
 *
 * Every SDK must behave the same way when the server sends a field or an enum value the
 * client does not know, and when a caller passes an enum value the SDK does not know: the
 * field is kept in `additionalProperties` and written back, the enum value stays the raw
 * string, and a caller's value is sent verbatim. The shared vector file is consumed by the
 * Go/Java/Python/TypeScript suites alike, so one SDK drifting is caught for all four.
 */
import * as fs from 'fs';
import * as path from 'path';
import { MultiFactorSignatureApi } from '../../src/internal/openapi/apis';
import {
  TgvalidatordGetBalancesReplyFromJSON,
  TgvalidatordGetBalancesReplyToJSON,
} from '../../src/internal/openapi/models/TgvalidatordGetBalancesReply';
import {
  TgvalidatordGetMultiFactorSignatureEntitiesInfoReplyFromJSON,
  TgvalidatordGetMultiFactorSignatureEntitiesInfoReplyToJSON,
} from '../../src/internal/openapi/models/TgvalidatordGetMultiFactorSignatureEntitiesInfoReply';
import {
  instanceOfTgvalidatordMultiFactorSignaturesEntityType,
  TgvalidatordMultiFactorSignaturesEntityTypeFromJSON,
} from '../../src/internal/openapi/models/TgvalidatordMultiFactorSignaturesEntityType';
import {
  TgvalidatordStakeAccountFromJSON,
  TgvalidatordStakeAccountToJSON,
} from '../../src/internal/openapi/models/TgvalidatordStakeAccount';
import {
  instanceOfTgvalidatordStakeAccountType,
  TgvalidatordStakeAccountTypeFromJSON,
} from '../../src/internal/openapi/models/TgvalidatordStakeAccountType';
import {
  instanceOfTgvalidatordTokenType,
  TgvalidatordTokenTypeFromJSON,
} from '../../src/internal/openapi/models/TgvalidatordTokenType';
import type { AssetAddressTypeV2 } from '../../src/models/asset';
import type { MultiFactorSignatureEntityType } from '../../src/models/multi-factor-signature';
import { MultiFactorSignatureService } from '../../src/services/multi-factor-signature-service';
import { pagedServices, RecordingTransport } from './pagination/harness';

const VECTORS_PATH = path.resolve(
  __dirname,
  '../../../scripts/resources/decode-tolerance-vectors.json'
);

interface ModelVector {
  name: string;
  kind: 'model';
  model: string;
  wire: Record<string, unknown>;
  expected: { lossless: boolean; rootAdditionalProperties: Record<string, unknown> };
}

interface EnumVector {
  name: string;
  kind: 'enum';
  enum: string;
  wire: string;
  expected: { value: string; known: boolean };
}

interface ServiceVector {
  name: string;
  kind: 'service';
  operation: string;
  request: Record<string, unknown>;
  reply: unknown;
  expected: Record<string, unknown>;
}

type Vector = ModelVector | EnumVector | ServiceVector;

const VECTORS: Vector[] = JSON.parse(fs.readFileSync(VECTORS_PATH, 'utf-8')) as Vector[];

const MODELS: Record<string, { fromJSON: (json: any) => any; toJSON: (value: any) => any }> = {
  tgvalidatordGetBalancesReply: {
    fromJSON: TgvalidatordGetBalancesReplyFromJSON,
    toJSON: TgvalidatordGetBalancesReplyToJSON,
  },
  tgvalidatordStakeAccount: {
    fromJSON: TgvalidatordStakeAccountFromJSON,
    toJSON: TgvalidatordStakeAccountToJSON,
  },
  tgvalidatordGetMultiFactorSignatureEntitiesInfoReply: {
    fromJSON: TgvalidatordGetMultiFactorSignatureEntitiesInfoReplyFromJSON,
    toJSON: TgvalidatordGetMultiFactorSignatureEntitiesInfoReplyToJSON,
  },
};

const ENUMS: Record<string, { fromJSON: (json: any) => unknown; isKnown: (value: any) => boolean }> =
  {
    tgvalidatordTokenType: {
      fromJSON: TgvalidatordTokenTypeFromJSON,
      isKnown: instanceOfTgvalidatordTokenType,
    },
    tgvalidatordMultiFactorSignaturesEntityType: {
      fromJSON: TgvalidatordMultiFactorSignaturesEntityTypeFromJSON,
      isKnown: instanceOfTgvalidatordMultiFactorSignaturesEntityType,
    },
    tgvalidatordStakeAccountType: {
      fromJSON: TgvalidatordStakeAccountTypeFromJSON,
      isKnown: instanceOfTgvalidatordStakeAccountType,
    },
  };

function mfsService(transport: RecordingTransport): MultiFactorSignatureService {
  return new MultiFactorSignatureService(new MultiFactorSignatureApi(transport.configuration));
}

const SERVICES: Record<string, (vector: ServiceVector) => Promise<void>> = {
  getMultiFactorSignatureInfo: async (vector) => {
    const transport = new RecordingTransport(vector.reply);

    const info = await mfsService(transport).get(vector.request.id as string);

    expect(info.entityType).toBe(vector.expected.entityType);
  },

  createMultiFactorSignatures: async (vector) => {
    const transport = new RecordingTransport(vector.reply);

    await mfsService(transport).create({
      entityType: vector.request.entityType as MultiFactorSignatureEntityType,
      entityIds: vector.request.entityIds as string[],
    });

    expect(transport.single().body).toEqual(expect.objectContaining(vector.expected.requestBody));
  },

  queryAssetAddresses: async (vector) => {
    const transport = new RecordingTransport(vector.reply);

    const page = await pagedServices(transport).assets.queryAssetAddresses(
      vector.request.assetId as string,
      { addressType: vector.request.addressType as AssetAddressTypeV2 }
    );

    expect(transport.single().body).toEqual(expect.objectContaining(vector.expected.requestBody));
    expect(
      page.items.map((row) => ({
        address: row.address,
        addressType: row.addressType,
        verified: row.verified,
      }))
    ).toEqual(vector.expected.rows);
    expect(page.excludedUnverified).toHaveLength(vector.expected.excludedCount as number);
  },
};

function byKind<K extends Vector['kind']>(kind: K): Array<Extract<Vector, { kind: K }>> {
  return VECTORS.filter((v): v is Extract<Vector, { kind: K }> => v.kind === kind);
}

describe('decode tolerance vectors', () => {
  it('loads the shared vectors file and knows every vector', () => {
    expect(VECTORS.length).toBeGreaterThan(0);
    for (const vector of VECTORS) {
      const known =
        vector.kind === 'model'
          ? vector.model in MODELS
          : vector.kind === 'enum'
            ? vector.enum in ENUMS
            : vector.operation in SERVICES;
      expect(`${vector.name}: ${known ? 'consumed' : 'NOT consumed'}`).toBe(
        `${vector.name}: consumed`
      );
    }
  });

  it.each(byKind('model').map((v) => [v.name, v] as const))(
    'model %s keeps unknown fields and re-serializes losslessly',
    (_name, vector) => {
      const codec = MODELS[vector.model];

      const decoded = codec.fromJSON(vector.wire);

      expect(decoded.additionalProperties).toEqual(vector.expected.rootAdditionalProperties);
      if (vector.expected.lossless) {
        expect(JSON.parse(JSON.stringify(codec.toJSON(decoded)))).toEqual(vector.wire);
      }
    }
  );

  it.each(byKind('enum').map((v) => [v.name, v] as const))(
    'enum %s keeps the raw value',
    (_name, vector) => {
      const codec = ENUMS[vector.enum];

      const decoded = codec.fromJSON(vector.wire);

      expect(decoded).toBe(vector.expected.value);
      expect(codec.isKnown(decoded)).toBe(vector.expected.known);
    }
  );

  it.each(byKind('service').map((v) => [v.name, v] as const))(
    'service %s',
    async (_name, vector) => {
      await SERVICES[vector.operation](vector);
    }
  );
});
